// @storybook-addons/docs/dist/register-with-configure.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var addons_1 = __importStar(require("@storybook/addons"));
var sync_external_url_1 = require("@storybook-addons/sync-external-url");
var constants_1 = require("./constants");
var docs_panel_1 = require("./components/docs-panel");
var DocsPanelEnhanced = sync_external_url_1.withFireEvents(docs_panel_1.DocsPanel);
exports.registerWithConfigure = function (_a) {
    var _b = (_a === void 0 ? {} : _a).tabTitle, tabTitle = _b === void 0 ? 'Documentation' : _b;
    addons_1.default.register(constants_1.ADDON_ID, function (api) {
        addons_1.default.add(constants_1.PANEL_ID, {
            type: addons_1.types.TAB,
            title: tabTitle,
            route: function (_a) {
                var storyId = _a.storyId;
                return "/docsx/" + storyId;
            },
            match: function (_a) {
                var viewMode = _a.viewMode;
                return viewMode === 'docsx';
            },
            render: function (_a) {
                var active = _a.active;
                return react_1.default.createElement(DocsPanelEnhanced, { api: api, active: active });
            },
        });
    });
};
