// @storybook-addons/docs/dist/components/docs-panel.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var styled_1 = __importDefault(require("@emotion/styled"));
var with_html_1 = __importDefault(require("react-markdown/with-html"));
var process_markdown_heading_1 = require("../lib/process-markdown-heading");
var constants_1 = require("../constants");
var params_1 = require("../params");
var skeleton_content_1 = require("./skeleton-content");
var empty_content_1 = require("./empty-content");
var typo_1 = require("./typo");
var theme_1 = require("./theme");
var markdown_renderers_1 = require("./markdown-renderers");
var Markdown = styled_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  font-family: var(--font-family);\n  font-size: var(--size-text-base);\n  line-height: var(--line-height-text-m);\n  background-color: var(--color-bg-default);\n  color: var(--color-typo-primary);\n"], ["\n  ", "\n  font-family: var(--font-family);\n  font-size: var(--size-text-base);\n  line-height: var(--line-height-text-m);\n  background-color: var(--color-bg-default);\n  color: var(--color-typo-primary);\n"])), theme_1.theme);
var Wrapper = styled_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  display: flex;\n  align-items: flex-start;\n  width: 960px;\n  margin: 0 auto;\n  padding: 24px 16px 64px;\n"], ["\n  display: flex;\n  align-items: flex-start;\n  width: 960px;\n  margin: 0 auto;\n  padding: 24px 16px 64px;\n"])));
var Content = styled_1.default.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n  width: 720px;\n  flex: 1 0 auto;\n\n  > :first-child {\n    margin-top: 0;\n  }\n\n  > :last-child {\n    margin-bottom: 0;\n  }\n"], ["\n  ", "\n  width: 720px;\n  flex: 1 0 auto;\n\n  > :first-child {\n    margin-top: 0;\n  }\n\n  > :last-child {\n    margin-bottom: 0;\n  }\n"])), typo_1.typo);
var Navigation = styled_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  margin-left: var(--space-xl);\n  flex: 1 1 240px;\n  position: sticky;\n  top: 0;\n"], ["\n  margin-left: var(--space-xl);\n  flex: 1 1 240px;\n  position: sticky;\n  top: 0;\n"])));
var NavigationList = styled_1.default.ul(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  padding-left: var(--space-xl);\n  border-left: 1px solid var(--color-bg-border);\n  list-style: none;\n  font-size: var(--size-text-m);\n  line-height: var(--line-height-text-m);\n"], ["\n  padding-left: var(--space-xl);\n  border-left: 1px solid var(--color-bg-border);\n  list-style: none;\n  font-size: var(--size-text-m);\n  line-height: var(--line-height-text-m);\n"])));
var NavigationItem = styled_1.default.li(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  margin: ", ";\n"], ["\n  margin: ", ";\n"])), function (props) { return (props.level === 2 ? "12px 0" : "8px 0 8px 12px"); });
var NavigationLink = styled_1.default.a(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  color: ", ";\n  text-decoration: none;\n  transition: color 100ms ease-in-out;\n  font-weight: ", ";\n\n  &:hover {\n    color: var(--color-link-external-hover);\n  }\n\n  &:focus {\n    outline: 2px solid var(--color-typo-primary);\n  }\n"], ["\n  color: ",
    ";\n  text-decoration: none;\n  transition: color 100ms ease-in-out;\n  font-weight: ", ";\n\n  &:hover {\n    color: var(--color-link-external-hover);\n  }\n\n  &:focus {\n    outline: 2px solid var(--color-typo-primary);\n  }\n"])), function (props) {
    return props.level === 2 ? 'var(--color-typo-primary)' : 'var(--color-link-external)';
}, function (props) { return (props.level === 2 ? 'var(--weight-bold)' : 'var(--weight-normal)'); });
exports.DocsPanelView = function (_a) {
    var api = _a.api, active = _a.active;
    var currentStoryData = api.getCurrentStoryData() || {};
    var userParams = api.getParameters(currentStoryData.id, params_1.PARAM_KEY) || {};
    var enableNavigation = __assign({}, params_1.defaultParams, userParams).enableNavigation;
    var isFirstRender = react_1.useRef(true);
    var _b = __read(react_1.useState(true), 2), shownSkeleton = _b[0], setShownSkeleton = _b[1];
    var _c = __read(react_1.useState({
        content: undefined,
        navigation: [],
    }), 2), _d = _c[0], content = _d.content, navigation = _d.navigation, setContent = _c[1];
    react_1.useEffect(function () {
        if (window.location.hash !== '') {
            var hash = decodeURIComponent(window.location.hash);
            var element = document.querySelector(hash);
            if (element !== null) {
                element.scrollIntoView();
            }
        }
    });
    react_1.useEffect(function () {
        var onAddReadme = function (_a) {
            var markdown = _a.content;
            var links = [];
            var processedContent = process_markdown_heading_1.processMarkdownHeading({
                markdown: markdown,
                onVisit: function (_a) {
                    var text = _a.text, link = __rest(_a, ["text"]);
                    return links.push(__assign({}, link, { text: text.replace(/`/g, '') }));
                },
            });
            if (content !== processedContent) {
                setContent({
                    content: processedContent,
                    // eslint-disable-next-line no-magic-numbers
                    navigation: links.filter(function (link) { return link.level > 1 && link.level < 4; }),
                });
            }
        };
        api.on(constants_1.ADD_README, onAddReadme);
        return function () {
            api.off(constants_1.ADD_README, onAddReadme);
        };
    }, [content, api]);
    react_1.useEffect(function () {
        if (isFirstRender.current) {
            isFirstRender.current = false;
        }
        else {
            setShownSkeleton(Boolean(content));
        }
    }, [content]);
    if (!active) {
        return null;
    }
    if (!content) {
        return (react_1.default.createElement(Markdown, null,
            react_1.default.createElement(Wrapper, null,
                react_1.default.createElement(Content, null, shownSkeleton ? react_1.default.createElement(skeleton_content_1.SkeletonContent, null) : react_1.default.createElement(empty_content_1.EmptyContent, null)))));
    }
    return (react_1.default.createElement(Markdown, null,
        react_1.default.createElement(Wrapper, null,
            react_1.default.createElement(Content, null,
                react_1.default.createElement(with_html_1.default, { escapeHtml: false, renderers: markdown_renderers_1.markdownRenderers, source: content })),
            enableNavigation && (react_1.default.createElement(Navigation, null,
                react_1.default.createElement(NavigationList, null, navigation.map(function (link) { return (react_1.default.createElement(NavigationItem, { key: link.url, level: link.level },
                    react_1.default.createElement(NavigationLink, { level: link.level, href: link.url }, link.text))); })))))));
};
exports.DocsPanel = react_1.memo(exports.DocsPanelView);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7;
